### PR TITLE
opswat v2 integration

### PR DESCRIPTION
**Pull request by yuvalbenshalom**
since we do not have an instance for opswat yet, this is just a short fix.
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/16301

## Description
A few sentences describing the overall goals of the pull request's commits.

## Does it break backward compatibility?
   - No - version 2

## Must have
- [x] Tests  - no instance
- [ ] Documentation (with link to it)
- [x] Code Review

### CODE REVIEW

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    USE_SSL = not PARAMS.get('insecure', False)
+    API_KEY = PARAMS.get('api_key', '')
+    USE_CLOUD = PARAMS.get('cloud', False)
+    HIGH_THRESHOLD = int(PARAMS.get('highPercnt', 66))
+    LOW_THRESHOLD = int(PARAMS.get('lowPercnt', 34))
+
+
+
+
+    if not demisto.params()['proxy']:
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: use `handle_proxy()`

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+            headers['filename'] = file['file_name']
+            files = {file['file_name']:file['file']}
+        LOG(f'running request with url={url}\n\tfiles={files}')
+        if method.upper() == 'GET':
+            res = requests.get(url, verify=USE_SSL, headers=headers)
+        elif method.upper() == 'POST':
+            res = requests.post(url, verify=USE_SSL, files=files, headers=headers)
+        else:
+            return_error(f'Got unsupporthed http method: {method}')
+            return ''
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: remove this

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        status = res.status_code
+        if status != 200:
+            if str(status) in HTTP_ERROR_CODES:
+                return_error(HTTP_ERROR_CODES[str(status)])
+                return ''
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: remove this

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+            else:
+                return_error(f'Request failed got status {status}')
+                return ''
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: remove this

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+        else:
+            return_error(f'Got unsupporthed http method: {method}')
+            return ''
+        status = res.status_code
+        if status != 200:
+            if str(status) in HTTP_ERROR_CODES:
+                return_error(HTTP_ERROR_CODES[str(status)])
+                return ''
+            else:
+                return_error(f'Request failed got status {status}')
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: can't we return here also `res.reason` or `res.content` ?

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+                return ''
+        if parse_json:
+            return res.json()
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: `.json()` might fail and raise `JSON could be decoded`, so best is to wrap in try except

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+        files = None
+        url = BASE_URL + url_suffix
+        headers = DEFAULT_HEADERS
+        if USE_CLOUD:
+            headers['apikey'] = API_KEY
+        if file:
+            headers['filename'] = file['file_name']
+            files = {file['file_name']:file['file']}
+        LOG(f'running request with url={url}\n\tfiles={files}')
+        if method.upper() == 'GET':
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: do we really need to split this to methods? can't we pass to `GET` request `files=None` ?

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+            return res.json()
+        else:
+            return res.content
+
+    def scan_file(file_entry_id):
+        try:
+            file_entry = demisto.getFilePath(file_entry_id)
+        except ValueError as e:
+            return_error(f'Failed to find file entry with id:\"{file_entry_id}\". got error: {e}')
+            return ''
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: remove this

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+        file_name = file_entry['name']
+        shutil.copy(file_entry['path'], file_name)
+        try:
+            with open(file_name, 'rb') as f:
+                res = http_req(method='POST',url_suffix='file',file={'file_name': file_name, 'file':f})
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: add whitespaces after commas and colon

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+        finally:
+            shutil.rmtree(file_name, ignore_errors=True)
+        return res, file_name
+
+
+    def get_hash_info(file_hash):
+        res = http_req(method='GET',url_suffix=f'hash/{file_hash}')
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: add whitespace after comma

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+        return res
+
+    def get_scan_result(scan_id):
+        res = http_req(method='GET',url_suffix=f'file/{scan_id}')
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: add whitespace after comma

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+        return res
+
+    def get_hash_info_command():
+        file_hash = demisto.args()['hash']
+        res = get_hash_info(file_hash)
+        if len(res) == 0:
+            return
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: we dont want to return anything in this case? like no results found or some error ?

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+        ec = {}
+        md ='# OPSWAT-Metadefender\n'
+        dbotScore = 0;
+
+        if 'file_info' in res:
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: use `.get()` for all key access in dict

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+            md += f'Detected AV: {total_detected_avs}/{total_avs}\n';
+            md += 'AV Name|Def Time|Threat Name Found\n';
+            md += '---|---|---\n';
+
+            avRes = scan_results['scan_details']
+            for key, scan in avRes.items():
+                def_time = scan['def_time']
+                threat_found = scan['threat_found']
+                md += f'{key}|{def_time}|{threat_found}\n'
+            percntBad = (total_detected_avs / total_avs) * 100
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: need to make sure these are integers

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+            file_type_description = file_info['file_type_description']
+            scan_all_result_a = scan_results['scan_all_result_a']
+            total_detected_avs = scan_results['total_detected_avs']
+            total_avs = scan_results['total_avs']
+
+            md += f'File name: {display_name}\n';
+            md += f'File description: {file_type_description}\n';
+            md += f'Scan result: {scan_all_result_a}\n';
+            md += f'Detected AV: {total_detected_avs}/{total_avs}\n';
+            md += 'AV Name|Def Time|Threat Name Found\n';
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: why not use tabletomarkdown?

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+                dbotScore = 3
+                ec[outputPaths['file']] = {
+                    'Hash': file_hash,
+                    'Malicious': {'Vendor': 'OPSWAT', 'Description': 'Result from OPSWAT-Metadefender'} }
+            elif percntBad >= LOW_THRESHOLD:
+                dbotScore = 2
+            else:
+                dbotScore = 1
+
+            ec['DBotScore'] = {'Indicator': file_hash, 'Type': 'hash', 'Vendor': 'OPSWAT', 'Score': dbotScore}
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: i think `Vendor` should be the integration `id`

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+            md += f'No results for hash {file_hash}\n'
+        entry = {
+                'ContentsFormat': formats['json'],
+                'Type': entryTypes['note'],
+                'Contents': res,
+                'ReadableContentsFormat': formats['markdown'],
+                'HumanReadable': md,
+                'EntryContext': ec
+                }
+        demisto.results(entry)
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: use `return_outputs`

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+
+
+    def scan_file_command():
+        file_entry_id = demisto.args()['fileId']
+        res, file_name= scan_file(file_entry_id)
+        if len(res) == 0:
+            return
+        scan_id = res['data_id']
+        md = '# OPSWAT-Metadefender\n'
+        ec = {}
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: remove this, just initialise the `ec` with `OPSWAT` key as in line 207

File: Integrations/integration-opswat-metadefender.yml [hunk 19]
```diff
@@ -304,7 +304,7 @@ script:
       required: true
       default: true
       description: File hash (Can be any hash type)
-    description: Check file hash on OPSWAT
+    description: Check file hash on OPSWAT (Deprecated)
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender.yml]: need also to deprecate the commands

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 20]
```diff
@@ -0,0 +1,342 @@
...
+            demisto.results('ok')
+        if demisto.command() == 'opswat-scan-file':
+            scan_file_command()
+        if demisto.command() == 'opswat-hash':
+            get_hash_info_command()
+        if demisto.command() == 'opswat-scan-result':
+            get_scan_result_command()
+    except Exception as e:
+        message = f'Unexpected error: {e}, traceback: {traceback.print_exc()}'
+        LOG(message)
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: `return_error` already does `LOG(<error message>)` so you can remove this

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 21]
```diff
@@ -0,0 +1,342 @@
...
+  name: proxy
+  defaultvalue: "false"
+  type: 8
+  required: false
+- display: The high threshold
+  name: highPercnt
+  defaultvalue: "66"
+  type: 0
+  required: false
+- display: 'The low threshold '
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: remove whitespace

File: Integrations/integration-opswat-metadefender-v2.yml [hunk 22]
```diff
@@ -0,0 +1,342 @@
...
+  name: api_key
+  defaultvalue: ""
+  type: 4
+  required: false
+- display: Cloud based
+  name: cloud
+  defaultvalue: "false"
+  type: 8
+  required: false
+- display: Trust any certificate (unsecure)
```
Comment (Itay4) [inline Integrations/integration-opswat-metadefender-v2.yml]: `unsecure` -> `not secure`

File: Integrations/integration-opswat-metadefender-v2.yml (hunk 5 above)
Comment (yuvalbenshalom) [reply Integrations/integration-opswat-metadefender-v2.yml]: there is nothin there but the status code number in this product

**Comment by yuvalbenshalom** [comment]
did some fixes but we will not have time for Alex to review because tomorrow we will not be Abe to test it.
just did the tests manually after the fixes.

File: Integrations/integration-opswat-metadefender.yml (hunk 19 above)
Comment (yuvalbenshalom) [reply Integrations/integration-opswat-metadefender.yml]: I did

**Pull request merged**